components/add-layers/shp/add-layers-shp.component.ts

```diff
@@ -69,27 +72,30 @@ export class HsAddLayersShpComponent implements OnInit {
   }
 
   /**
-   * @param endpoint
-   * @param layerName
+   * @param endpoint Selected endpoint (should be Layman)
+   * @param layerName Name of the layer to describe
+   * @returns {Promise} Description of Layman layer
    */
-  describeNewLayer(endpoint, layerName): Promise<any> {
-    return new Promise((resolve, reject) => {
-      this.hsLaymanService
-        .describeLayer(endpoint, layerName)
-        .then((descriptor) => {
-          if (
-            ['STARTED', 'PENDING', 'SUCCESS'].includes(descriptor.wms.status)
-          ) {
-            setTimeout(() => {
-              this.describeNewLayer(endpoint, layerName).then((response) =>
-                resolve(response)
-              );
-            }, 2000);
-          } else {
-            resolve(descriptor);
-          }
-        });
-    });
+  async describeNewLayer(
+    endpoint: HsEndpoint,
+    layerName: string
+  ): Promise<HsLaymanLayerDescriptor> {
+    try {
+      const descriptor = await this.hsLaymanService.describeLayer(
+        endpoint,
+        layerName
+      );
+      if (['STARTED', 'PENDING', 'SUCCESS'].includes(descriptor.wms.status)) {
+        setTimeout(async () => {
+          return await this.describeNewLayer(endpoint, layerName);
```

Comment:
I was struggling with this timeout and thats the reason why I havent rewriotten to promises. It was returning inside setTimeout instead of describeNewLayer. I suspect it might be that here too. The way to check would be to upload a large shp file which takes time to process.

Reply:
You are commenting on an outdated file, are you aware of that? Exactly this part was the problem – I had to promisify the `setTimeout()` with `new Promise()` and resolve from inside of it.